test/framework/src/main/java/org/elasticsearch/test/TransportVersionUtils.java

```diff
@@ -128,11 +122,13 @@ public static TransportVersion randomPreviousCompatibleVersion(Random random, Tr
 
     /** returns the first future compatible version */
     public static TransportVersion compatibleFutureVersion(TransportVersion version) {
-        final Optional<TransportVersion> opt = ALL_VERSIONS.stream()
-            .filter(version::before)
-            .filter(v -> isCompatible(v, version))
-            .findAny();
-        assert opt.isPresent() : "no future compatible version for " + version;
-        return opt.get();
+        int versionIdx = Collections.binarySearch(ALL_VERSIONS, version);
```

Comment:
Instead of starting the binary search at index 0, you could specify the bounds to search to include minCompatIdx so that the Math.max is not needed after?

Reply:
I've just realised I changed how this method behaves when I refactored it. It's only used in one place, I'll see if I can do something else instead

Reply:
I'll update this method in a separate PR

Reply:
#94366